vernac/declareObl.ml

```diff
@@ -531,7 +531,12 @@ let obligation_terminator entries uctx { name; num; auto } =
       if prg.prg_poly then uctx
       else UState.union prg.prg_ctx uctx
     in
-    let univs = UState.univ_entry ~poly:prg.prg_poly uctx in
+    (* XXX: use the combined to_constr_and_evars form when merged *)
+    let vars = Vars.universes_of_constr body in
+    let vars = Option.cata
+        (fun ty -> Univ.LSet.union (Vars.universes_of_constr ty) vars) vars ty in
+    let uctx = UState.restrict uctx vars in
+    let univs = UState.check_univ_decl ~poly:prg.prg_poly uctx prg.prg_univdecl in
```

Comment:
Basically the problem is described in the comment a bit below:
~~~
        (* The first obligation, if defined,
           declares the univs of the constant,
           each subsequent obligation declares its own additional
           universes and constraints if any *)
~~~
if we restrict then this isn't true anymore.

Reply:
Thanks for pointing out the issue; I don't know enough about univs to understand what you mean, trying to figure out. Do you any idea on how to move forward with the PR meanwhile?

Reply:
We can do just the first commit for now (assuming it works)
Long term I think for monomorphic constants we may want to declare the type universes upfront, ie
~~~
Lemma foo : Type@{u}. (* declare_univ u here, before taking the snapshot for resetting to at qed *)
Proof. (* bla bla *) Qed. (* no need to declare u here *)
~~~
This isn't just for obligations, I remember thinking it made sense for side effects too, I forgot the details though

Reply:
Oh good point about the snapshot; I planned to change that actually; I think I'm gonna indeed proceed with that change first then revisit this.